compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeNarrower.java

```diff
@@ -221,20 +223,33 @@ public SymbolEnv evaluateTruth(BLangExpression expr, BLangNode targetNode, Symbo
      * @param expr Expression to evaluate
      * @param targetNode node to which the type narrowing applies
      * @param env Current environment
+     * @param isBinaryExpr Indicates whether the current context is a binary expression
      * @return target environment
      */
-    public SymbolEnv evaluateFalsity(BLangExpression expr, BLangNode targetNode, SymbolEnv env) {
+    public SymbolEnv evaluateFalsity(BLangExpression expr, BLangNode targetNode, SymbolEnv env, boolean isBinaryExpr) {
         Map<BVarSymbol, NarrowedTypes> narrowedTypes = getNarrowedTypes(expr, env);
         if (narrowedTypes.isEmpty()) {
             return env;
         }
 
         SymbolEnv targetEnv = getTargetEnv(targetNode, env);
-        narrowedTypes.forEach((symbol, typeInfo) -> {
-            BVarSymbol originalSym = getOriginalVarSymbol(symbol);
-            symbolEnter.defineTypeNarrowedSymbol(expr.pos, targetEnv, originalSym, typeInfo.falseType,
+        for (Map.Entry<BVarSymbol, NarrowedTypes> narrowedType : narrowedTypes.entrySet()) {
+            BType falseType = narrowedType.getValue().falseType;
+            BType trueType = narrowedType.getValue().trueType;
+
+            BVarSymbol originalSym = getOriginalVarSymbol(narrowedType.getKey());
+
+            if (isBinaryExpr) {
+                falseType = falseType == symTable.semanticError ?
+                        types.getRemainingType(originalSym.type, trueType) : falseType;
```

Comment:
`types.getRemainingType(originalSym.type, trueType)` seems to return the type itself if both the arguments are the same type. Shouldn't be semanticError? Or `never`? Shall we create an issue?

Reply:
Created https://github.com/ballerina-platform/ballerina-lang/issues/33567